Improve error on conflict for nix profile install

bobvanderlinden opened the pull request:
# Motivation

Currently `nix profile install` gives a confusing and unhelpful error message whenever a package is installed that has conflicting files. This often happens when installing a different version of the same package.

This PR is an attempt to improve the experience in such situations to some extend. I see this as a small improvement, but it doesn't really reach the end-goal of a good UX yet.

As an example, this is what happens before this PR:

```console
$ nix profile install github:cachix/devenv/v0.5
$ nix profile install github:cachix/devenv/v0.4
error: files '/nix/store/k0gwhiwfs8c218kph78vx75bph51iw31-devenv/bin/devenv' and '/nix/store/7i4hxhw44lr7r64zlmp4zq7cqpsx7smw-devenv/bin/devenv' have the same priority 5; use 'nix-env --set-flag priority NUMBER INSTALLED_PKGNAME' or type 'nix profile install --help' if using 'nix profile' to find out how to change the priority of one of the conflicting packages (0 being the highest priority)
```

This is what happens after this PR:

```console
$ nix profile install github:cachix/devenv/v0.5
$ nix profile install github:cachix/devenv/v0.4
error: There is an existing package that has a conflicting file:

         /nix/store/7i4hxhw44lr7r64zlmp4zq7cqpsx7smw-devenv/bin/devenv

       The conflicting file:

         /nix/store/k0gwhiwfs8c218kph78vx75bph51iw31-devenv/bin/devenv

       You can remove the existing package using:

         nix profile remove github:cachix/devenv/v0.4

       You can also install the new package next to the existing package by assigning a priority to the package.
       The conflicting packages have a priority of 5.
       You can prioritize the new package using:

         nix profile install github:cachix/devenv/v0.4 --priority 4

       You can prioritize the existing package using:

         nix profile install github:cachix/devenv/v0.4 --priority 6
```

The new changes allows `nix profile install` to give more context.

This also avoids mentions of `nix-env` and `nix profile` when using `builtins.buildEnv`. Since `buildEnv` can be used in other places (unrelated to the Nix profile), it makes sense to avoid mentioning `nix-env` and `nix profile`, as they likely do not apply in these cases.

# Context

Related to https://github.com/NixOS/nix/issues/7530

See [my suggestion](https://github.com/NixOS/nix/issues/7530#issuecomment-1419123524) for a good UX on `nix profile`s. The current PR is merely an improvement for the error message on conflicts.

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

domenkozar commented:
This looks great to me in terms of the error message improvement, but I'll let those that know the codebase comment on the C++ bits.

bobvanderlinden commented:
I made some more changes so that the `INSTALLABLE` of the suggested `nix profile install` commands can be filled in. The implementation first throws an error containing the conflicting files. Then the package path (`/nix/store/XXXX-package`) is looked up for the conflicting files. Lastly the INSTALLABLE from the manifest is found using the package path. All of this information is used in the resulting error.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/contributing-to-nix/25605/1

asymmetric commented on src/nix/profile.cc:
```diff
@@ -329,7 +335,63 @@ struct CmdProfileInstall : InstallablesCommand, MixDefaultProfile
+            }();
+
+            throw Error(
+                "There is an existing package that has a conflicting file:\n"
+                "\n"
+                "  %1%\n"
+                "\n"
+                "The conflicting file:\n"
+                "\n"
+                "  %3%\n"
```
This reads confusing to me, as if the second store path is contained in the first.

```suggestion
                "An existing package already symlinks the following file to the user's profile:\n"
                "\n"
                "  %1%\n"
                "\n"
                "The conflicting file:\n"
                "\n"
                "  %3%\n"
```

bobvanderlinden replied:
Indeed, it's worded a bit convoluted. I'd like to avoid mentioning symlinking to the user. They don't necessarily know profiles are build from symlinks. I reworded it as:

> An existing package already provides the following file to the user's profile

Ericson2314 commented on src/libstore/path.hh:
```diff
@@ -62,6 +63,15 @@ public:
     static StorePath dummy;
 
     static StorePath random(std::string_view name);
+
+    /**
+     * Parses absolute store paths, like
+     * "/nix/store/XXXXXX-name", to a StorePath.
+     * Throws BadStorePath when the path cannot be parsed
+     * as a StorePath.
+     */
+    static StorePath parseAbsolutePath(Path path);
+    static std::optional<StorePath> maybeParseAbsolutePath(Path path);
```
We don't want these, we want to use the parsing methods on `Store` instead.

Ericson2314 replied:
```suggestion
```

Ericson2314 commented on src/libstore/builtins/buildenv.hh:
```diff
@@ -12,6 +12,66 @@ struct Package {
+        const Path conflictingFile,
+        int priority
+    )
+        : Error(
+            "Unable to build profile. There is a conflict for the following files:\n"
+            "\n"
+            "  %1%\n"
+            "  %2%",
+            originalFile,
+            conflictingFile
+        )
```
Side note, we really have a separate virtual method for *rendering* errors so we aren't storing both the structured data and a big long preformatted string. This is an awkard redundancy that just tempts people into overusing the plain `Error` type.

Ericson2314 replied:
Side note split into issue #7865.

bobvanderlinden replied:
Hmm, I see I could override `what`, but quite a few places use `.msg()` instead of `what`. `msg` doesn't seem virtual.

I looked around in the codebase, but it seems there aren't that many custom (structured) errors, as they all use `MakeError`.

What would be the right way to implement 'lazy' error messages?

Ericson2314 replied:
@bobvanderlinden to be clear, I didn't mean this to block this PR. But if you would like to take a stab at this issue, I am happy to provide guidance over there.

Ericson2314 commented on src/nix/profile.cc:
```diff
@@ -329,7 +335,63 @@ struct CmdProfileInstall : InstallablesCommand, MixDefaultProfile
+        } catch (BuildProfileConflictError & conflictError) {
+            std::string fallbackInstallable = "INSTALLABLE";
+            std::string conflictingInstallable = [&]() {
+                auto maybeConflictingStorePath = StorePath::maybeParseAbsolutePath(conflictError.conflictingPkgPath);
```
```suggestion
                auto maybeConflictingStorePath = store->maybeParseStorePath(conflictError.conflictingPkgPath);
```

bobvanderlinden replied:
Oh, that is a lot more practical than introducing parseAbsolutePath :sweat_smile: Thanks :+1:

Ericson2314 commented on src/libstore/path.cc:
```diff
@@ -52,6 +52,23 @@ StorePath StorePath::random(std::string_view name)
     return StorePath(hash, name);
 }
 
+StorePath StorePath::parseAbsolutePath(Path path) {
+    // Find the installable that contains the conflicting package.
+    auto pathSeparatorIndex = path.rfind('/');
+    if (pathSeparatorIndex == std::string::npos) {
+        throw BadStorePath("path '%s' is not an absolute path", path);
+    }
+    return StorePath(path.substr(pathSeparatorIndex + 1));
+}
+
+std::optional<StorePath> StorePath::maybeParseAbsolutePath(Path path) {
+    try {
+        return parseAbsolutePath(path);
+    } catch (BadStorePath &) {
+        return {};
+    }
+}
+
```
```suggestion
```

Ericson2314 commented:
@bobvanderlinden fair warning I don't yet have perms to merge this, but I can still shepherd along making it an easier approval for those that do.

Ericson2314 commented:
@bobvanderlinden This is starting to look quite good! Thanks for doing this!

Correct me if I am wrong, but `BuildEnvFileConflictError` is just used in a private-to-the-file `static` function, so you should be able to *not* expose it in the header, but just have it in `buildenv.cc`.

Since it is for internal use only, and will always be caught, I think it would be also fine to just make it inherit `std::exception` rather than `Error`, and then you don't have to write a nice message which would be dead code.

bobvanderlinden commented:
> @bobvanderlinden This is starting to look quite good! Thanks for doing this!
> 
> Correct me if I am wrong, but `BuildEnvFileConflictError` is just used in a private-to-the-file `static` function, so you should be able to _not_ expose it in the header, but just have it in `buildenv.cc`.
> Since it is for internal use only, and will always be caught, I think it would be also fine to just make it inherit `std::exception` rather than `Error`, and then you don't have to write a nice message which would be dead code.

Ah you're right! That makes a lot of sense, thanks :+1: Updated the PR.

Ericson2314 commented on src/nix/profile.cc:
```diff
@@ -41,6 +41,12 @@ struct ProfileElement
     bool active = true;
     int priority = defaultPriority;
 
+    bool hasStorePath(const StorePath & basePath) const {
+        return any_of(storePaths.begin(), storePaths.end(), [&](const StorePath & storePath) {
+            return storePath == basePath;
+        });
+    }
```
I like the combinators! But try https://en.cppreference.com/w/cpp/container/set/contains or (if our C++20 is not new enough) https://en.cppreference.com/w/cpp/container/set/find

Ericson2314 replied:
Then you don't need this method

bobvanderlinden replied:
I missed this method in my search for a right candidate. That makes it quite a bit cleaner :+1:

Ericson2314 commented on src/libstore/builtins/buildenv.cc:
```diff
@@ -13,6 +13,24 @@ struct State
     unsigned long symlinks = 0;
 };
 
+class BuildEnvFileConflictError : public std::exception
```
```suggestion
/** This is private to this file and always caught, so we don't worry about it rendering in a nice fancy way etc. */
class BuildEnvFileConflictError : public std::exception
```

Ericson2314 replied:
And maybe line wrap it :)

bobvanderlinden replied:
I rewrote the comment a bit. Hope you like it ;)

Ericson2314 commented:
Thanks @bobvanderlinden, this is very close to being ready!

bobvanderlinden commented:
> Thanks @bobvanderlinden, this is very close to being ready!

Thanks for the quick reviews and suggestions. This improves my long forgotten C++ knowledge a bit too. :partying_face:

Ericson2314 reviewed (approved):
Looks good!

Ericson2314 commented:
Thanks so much, @bobvanderlinden, I've approved it and will recommend it for those with merge access.

roberth commented:
Not testing error messages leads to a lot of untested code, including this code. We should consider testing more error messages. They're important too.

asymmetric commented on src/nix/profile.cc:
```diff
@@ -329,7 +329,63 @@ struct CmdProfileInstall : InstallablesCommand, MixDefaultProfile
+                "\n"
+                "  %1%\n"
+                "\n"
+                "This file conflicts with a file from the new package:\n"
```
Here it's not 100% clear if "this" is referring to the one below or the one above, so I would reword as:

```suggestion
                "This is the conflicting file from the new package:\n"
```

Ericson2314 commented:
@roberth Agreed, see `src/libutil/tests/logging.cc` for an example of *commented out* :scream: testing of diagnostics. If we put the expected output in separate file (see "golden testing") these escaped string literals, it might be more maintainable.

roberth commented:
> @roberth Agreed, see `src/libutil/tests/logging.cc` for an example of _commented out_ scream testing of diagnostics. If we put the expected output in separate file (see "golden testing") these escaped string literals, it might be more maintainable.

Crucially this needs a mode to overwrite the outputs, to fix the maintenance problem, turning tedious work into a skimmable diff review.

fricklerhandwerk commented:
Discussed in the Nix team meeting:

- @thufschmitt: this is a great improvement overall
- @edolstra: philosophical objection to putting manual-style content into error messages
  - better to have short description linking to the manual that tells how to fix something
  - @fricklerhandwerk: had this discussion in the past, currently don't have a good means of linking to the manual.
    - having versioned online documentation would make that easier, and should be a priority for @NixOS/documentation-team
   - then again, since the error message is adapted to the actual situation, this is something the manual can't do and therefore this change is fine
   - ideally we'd make this a markdown message and run a formatter on them for better presentation
- assigned to @fricklerhandwerk for reviewing the user-facing part

edolstra commented on src/libstore/builtins/buildenv.cc:
```diff
@@ -13,6 +13,30 @@ struct State
+ * It inherits from std::exception instead of nix::Error to
+ * avoid having to render pretty errors.
+ */
+class BuildEnvFileConflictError : public std::exception
```
Why is this a subclass of `std::exception` instead of `Error`? In most of the Nix code base we catch `Error`, not `std::exception`.

bobvanderlinden replied:
See https://github.com/NixOS/nix/pull/7788#issuecomment-1438696402

edolstra commented on src/libstore/builtins/buildenv.cc:
```diff
@@ -13,6 +13,30 @@ struct State
+class BuildEnvFileConflictError : public std::exception
+{
+public:
+    const Path originalFile;
+    const Path conflictingFile;
+    int priority;
+
+    BuildEnvFileConflictError(
+        const Path originalFile,
+        const Path conflictingFile,
+        int priority
+    )
+        : originalFile(originalFile)
+        , conflictingFile(conflictingFile)
+        , priority(priority)
+    {}
```
It would be nicer to use designated initializers, e.g.
```
BuildEnvFileConflictError {
    .originalFile = ...,
    .conflictingFile = ...,
    .priority = ...
}
```
which is more readable and removes the need for a constructor.

bobvanderlinden replied:
Hmm, I looked into this, is this even possible when it is inheriting from std::exception?

Not entirely sure, but it seems it is mostly possible for structs?

```
src/libstore/builtins/buildenv.cc:113:25: error: designated initializers cannot be used with a non-aggregate type ‘nix::BuildEnvFileConflictError’
```

Ericson2314 replied:
Ah maybe `std::exception` isn't really needed, and then this can be a simple plain old struct.

bobvanderlinden replied:
Aah, yes, without using `std::exception` I could indeed change it to struct, which works for throwing :+1:

edolstra commented on src/libstore/builtins/buildenv.hh:
```diff
@@ -12,6 +12,40 @@ struct Package {
+        const Path conflictingFile,
+        const Path conflictingPkgPath,
+        int priority
+    )
```
Idem, use designated initializers.

bobvanderlinden replied:
By using struct this works. Incorporated into the PR :+1: Thanks for the suggestion, named-parameters-like is much clearer to read.

edolstra commented on src/nix/profile.cc:
```diff
@@ -329,7 +329,63 @@ struct CmdProfileInstall : InstallablesCommand, MixDefaultProfile
+
+            throw Error(
+                "An existing package already provides the following file to the user's profile:\n"
+                "\n"
```
Probably nicer to write this using multiline string literals, i.e. `R"(...)"`.

bobvanderlinden replied:
I have looked into various ways of doing multiline strings in C++. I think R"(...)" was one of them and it seemed it didn't support removing indentation, so it required me to put all text on the first column. The current solution seemed like one that did properly indent, which looked a bit better.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2023-02-24-nix-team-meeting-minutes-35/25757/1

bobvanderlinden commented:
> Not testing error messages leads to a lot of untested code, including this code. We should consider testing more error messages. They're important too.

Fully agree. I really like snapshot testing for these kinds of cases with large pieces of unstructured text. I have added a case to `tests/nix-profile.sh`. It's an example, but I don't know whether this is the way we want to test features like these. For now it works, but when something changes it might be useful to be able to update the snapshot automatically.

Maybe eventually look into a more reusable way to do snapshots? And/or combine it with a testing framework like [bats](https://github.com/bats-core/bats-core)?

Let me know whether the current test is wanted for this PR.

Ericson2314 commented on tests/nix-profile.sh:
```diff
@@ -140,6 +140,32 @@ printf World2 > $flake2Dir/who
+diff -u <(nix profile install $flake2Dir 2>&1 1> /dev/null || true) <(cat << EOF
+error: An existing package already provides the following file to the user's profile:
+
+         /tmp/nix-shell.dvSd2u/nix-test/tests/nix-profile/store/d4nwfvm9sz012b4i2331i2rrr5x4byfr-profile-test-4.0/bin/hello
```
You'll need to detail with the store dir changing (use `$NIX_STORE_DIR`) and the hash change (not so sure).

bobvanderlinden replied:
Yes, the hash is not reproducible on different machines with different test roots. I'm looking for a way to find the store path. I had `nix eval ${flake1Dir}"#default.outPath" --raw`, but it doesn't work yet. I'm going look into this at a later time.

bobvanderlinden replied:
Ended up using `nix build --no-link --print-out-paths ${flake2Dir}"#default.out"`.

fricklerhandwerk commented on src/nix/profile.cc:
```diff
@@ -329,7 +329,63 @@ struct CmdProfileInstall : InstallablesCommand, MixDefaultProfile
+                "\n"
+                "  nix profile remove %5%\n"
+                "\n"
+                "You can also install the new package next to the existing package by assigning a priority to the package.\n"
```
```suggestion
                "The new package can also be installed next to the existing one by assigning a different priority.\n"
```

I suppose it already has some priority, and the user would have to change it.

fricklerhandwerk commented on src/nix/profile.cc:
```diff
@@ -329,7 +329,63 @@ struct CmdProfileInstall : InstallablesCommand, MixDefaultProfile
+            }();
+
+            throw Error(
+                "An existing package already provides the following file to the user's profile:\n"
```
```suggestion
                "An existing package already provides the following file:\n"
```

@edolstra seemed a bit surprised over the verbosity of the output, and I also think we should try to keep visual noise to a minimum while offering all the assistance we can. That's what I'm trying to do in these suggestions, so don't worry about getting many of them.

fricklerhandwerk commented on src/nix/profile.cc:
```diff
@@ -329,7 +329,63 @@ struct CmdProfileInstall : InstallablesCommand, MixDefaultProfile
+                "\n"
+                "  %3%\n"
+                "\n"
+                "You can remove the existing package using:\n"
```
```suggestion
                "To remove the existing package:\n"
```

Saying "you can ... using" doesn't really add information. The user always can something using something. We may leave it out. I understand addressing the user may sound friendlier, but a wall of text may be overwhelming, and should only happen if all of that is relevant information.

fricklerhandwerk commented on src/nix/profile.cc:
```diff
@@ -329,7 +329,63 @@ struct CmdProfileInstall : InstallablesCommand, MixDefaultProfile
+                "\n"
+                "You can also install the new package next to the existing package by assigning a priority to the package.\n"
+                "The conflicting packages have a priority of %6%.\n"
+                "You can prioritize the new package using:\n"
```
```suggestion
                "To prioritise the new package:\n"
```

The docs team decided to use British English. We'll have to write it down somewhere...

fricklerhandwerk commented on src/nix/profile.cc:
```diff
@@ -329,7 +329,63 @@ struct CmdProfileInstall : InstallablesCommand, MixDefaultProfile
+                "\n"
+                "  nix profile install %5% --priority %7%\n"
+                "\n"
+                "You can prioritize the existing package using:\n"
```
```suggestion
                "To prioritise the existing package:\n"
```

fricklerhandwerk reviewed (approved):
In really like the informative error message, we can use lots more of that! Only suggestions on more concise wording from my side. Great work.

bobvanderlinden commented:
The test did show that I made a mistake with the `nix profile remove` command where the conflicting installable was mixed up with the original installable, which I have now fixed. Good idea to add this test. Glad I did.

Thanks you all for the suggestions. I've incorporated most of them and resolved the discussions. Quite a few improvements to the PR already :+1:

> ideally we'd make this a markdown message and run a formatter on them for better presentation

Is this something I can look into for this PR?

roberth commented:
> > ideally we'd make this a markdown message and run a formatter on them for better presentation
> 
> Is this something I can look into for this PR?

Better to do that in a followup, although it may make the testing story worse, as the layout becomes dependent on the length of the inputs.

bobvanderlinden commented:
I have reworked the implementation in this PR and left the tests as-is.

I have removed `BuildProfileConflictError` and opted to merely throw `BuildEnvFileConflictError`. `BuildEnvFileConflictError` now holds `fileA` and `fileB`, which represent the 2 files that were conflicting inside `buildEnv`. Since we do not know at that time whether the files were from the original package or the newly installed package, I have called them A and B.

Determining which file belongs to the original/new package is done entirely in `CmdProfileInstall::run`. Packages that are being installed are placed at the end of manifest.elements, so from this I look up any package that matches one of the conflicting file paths at the end of `manifest.elements` to find which the new package and its conflicting file. I do the same thing for the already installed package, but by looking at the beginning of `manifest.elements`.

This gives an implementation that doesn't run into sorting issues that I can into before. Bonus: the code is cleaner this way as well.

bobvanderlinden commented:
Hmm, MacOS in CI failed to build due to usage of C++20 ranges. I guess a different version of clang is used on MacOS to build Nix?
I now avoided using ranges and replaced it with `begin()` and `rbegin()` iterators. I tried compiling using clang and using `nix develop .#nix-gccStdenv` as well. :crossed_fingers:

fricklerhandwerk commented:
@Ericson2314 I'll unsubscribe since the user-facing part is good, please continue with shepherding the implementation.

Ericson2314 commented:
Implementation looks good to me!

fricklerhandwerk commented:
Merging it then.

bobvanderlinden commented:
Thank you all 🥳👍

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nix-team-report-2022-10-2023-03/27486/1
